xcdat/axis.py

```diff
@@ -0,0 +1,270 @@
+
+        :param bounds: Base latitude boundaries
+        :type bounds: np.ndarray
+        :return: Adjusted latitude boundaries
+        :rtype: np.ndarray
+        """
+        bounds[0, ...] = np.maximum(-90.0, np.minimum(90.0, bounds[0, ...]))
+        bounds[-1, ...] = np.maximum(-90.0, np.minimum(90.0, bounds[-1, ...]))
+
+        return bounds
+
+    def _adjust_lon_bounds(self, bounds: np.ndarray) -> np.ndarray:
+        """If necessary, adjusts longitude boundaries to avoid floating point errors.
+
+        The endpoints may also be adjusted to ensure circularity.
+
+        :param bounds: Base longitude boundaries
+        :type bounds: np.ndarray
+        :return: Adjusted longitude boundaries
+        :rtype: np.ndarray
+        """
+        # Example: [[-1, -0.5], [0.5, 1]]
+        # [[min_bound_left, min_bound_right][max_bound_left, max_bound_right]]
+        min_bound_left = bounds[0, 0]  # = -1
+        min_right_right = bounds[0, 1]  # = -0.5
+        max_bound_left = bounds[-1, 0]  # = 0.5
+        max_bound_right = bounds[-1, 1]  # = 1
+
+        # Variables are used to ensure circularity
+        max_degree_interval = 360.0
+        max_degree_threshold = np.minimum(
+            0.01, abs(min_right_right - min_bound_left) * 0.1
+        )
+
+        # Check if the bounds are within the max degree limit to determine if rounding
+        # might be necessary
+        bounds_diff = abs(max_bound_right - min_bound_left)
+        bounds_diff_with_max_degree = abs(bounds_diff - max_degree_interval)
+
+        # For example, for 0.001 < 0.01, the bounds are below the threshold and considered
+        # near the max degree interval so adjustment is necessary
+        bounds_near_max_degree: bool = (
+            bounds_diff_with_max_degree < max_degree_threshold
+        )
+
+        if bounds_near_max_degree:
+            min_bound_int_limit = abs(min_right_right - min_bound_left) * 0.01
+            min_left_val_near_int: bool = (
+                abs(min_bound_left - np.floor(min_bound_left + 0.5))
+                < min_bound_int_limit
+            )
+
+            max_bound_int_limit = abs(max_bound_right - max_bound_left) * 0.01
+            max_right_val_near_int: bool = (
+                abs(max_bound_right - np.floor(max_bound_right + 0.5))
+                < max_bound_int_limit
+            )
+
+            # For (-180, 180), if either bound is near an integer value, round both integers
+            # Otherwise it is not needed if all values are positive (0, 360)
+            if (
+                min_left_val_near_int or max_right_val_near_int
+            ) and min_bound_left * max_bound_right < 0:
+                bounds[0, 0] = np.floor(min_bound_left + 0.5)
+                bounds[0, 1] = np.floor(max_bound_right + 0.5)
+            else:
+                if max_bound_right > min_bound_left:
+                    bounds[-1, 1] = round(min_bound_left + max_degree_interval, 4)
```

Comment:
I had to add `round()` here or I cannot test the result since it can be < -0.0009999999

Reply:
This might explain why the bounds are exact in some datasets?

Reply:
Would it be better to move `round` out of the implementation and into the test? This might surprise the user if they have a longitude axis starting from a small number <4 decimals, not sure how common this is.

Example: This produces bounds where the last interval is not complete.
```python
ds = xr.Dataset(None, coords={"lon": xr.DataArray(np.arange(0.05125, 360, 0.5), attrs={"units": "degrees_west"})})
x = ds.axis.get_bounds("lon")
x[:,1]-x[:,0]
```

Reply:
Thanks for the suggestion. I agree, this is a special test case and it would be better to round in the test.

Reply:
Fixed by removing `round()` and using[ `numpy.allclose()`](https://numpy.org/doc/stable/reference/generated/numpy.allclose.html#numpy.allclose) to compare two float arrays